File: OpenBurnBarMobile/Services/RollbackService.swift
import FirebaseAuth
import FirebaseCore
import FirebaseFirestore
import Foundation
import Observation
import OpenBurnBarCore

// MARK: - Rollback Service (Hermes Square §6.10)
//
// Listens to `users/{uid}/cli_sessions/{sessionID}/snapshots` for the
// per-session snapshot index the Mac publishes. Submits a `RollbackRequest`
// to `users/{uid}/rollback_requests/{requestID}` which the Mac claims and
// applies.
//
// Privacy: the scope JSON (which can embed an absolute file path for a
// `singleFile` rollback) and the Mac-written resolution diagnostic are sealed
// with the per-user cloud vault key before they touch Firestore
// (`sealedScope` / `sealedErrorMessage`). Snapshot index fields the Mac writes
// (`actionLabel` / `touchedFiles` / `macSnapshotPath`) are likewise sealed.
// Every reader keeps a legacy plaintext fallback so in-flight / pre-migration
// documents still render. No Cloud Function reads these collections — they are
// pure store-and-forward between the user's own devices, so a vault seal (no
// keyed hash) is the complete fix.

@MainActor
@Observable
final class RollbackService {
    static let shared = RollbackService()

    private(set) var snapshotsBySession: [String: [RollbackSnapshot]] = [:]
    private(set) var pendingRequests: [RollbackRequest] = []
    private(set) var inlineError: String?

    private var snapshotObservations: [String: ListenerRegistration] = [:]
    private var requestObservation: ListenerRegistration?

    private let firestoreProvider: () -> Firestore

    init(firestoreProvider: @escaping () -> Firestore = { Firestore.firestore() }) {
        self.firestoreProvider = firestoreProvider
    }

    func startObservingSession(_ sessionID: String) {
        guard FirebaseApp.app() != nil, snapshotObservations[sessionID] == nil else { return }
        guard let uid = Auth.auth().currentUser?.uid else { return }
        let ref = firestoreProvider()
            .collection("users").document(uid)
            .collection("cli_sessions").document(sessionID)
            .collection("snapshots")
            .order(by: "sequence")
        let reg = ref.addSnapshotListener { [weak self] snapshot, _ in
            Task { @MainActor in
                guard let self else { return }
                let documents = snapshot?.documents ?? []
                // Resolve the vault key once per listener fire so sealed
                // snapshot fields can be opened; legacy plaintext docs still
                // decode when the key is nil.
                let vaultKey = try? await MobileCloudVaultKeyAccess.keyForReading(
                    uid: uid,
                    firestore: self.firestoreProvider()
                )?.keyData
                let parsed: [RollbackSnapshot] = documents.compactMap { doc in
                    Self.decodeSnapshot(
                        data: doc.data() as NSDictionary,
                        documentID: doc.documentID,
                        sessionID: sessionID,
                        vaultKey: vaultKey
                    )
                }
                self.snapshotsBySession[sessionID] = parsed
            }
        }
        snapshotObservations[sessionID] = reg
    }

    func stopObservingSession(_ sessionID: String) {
        snapshotObservations[sessionID]?.remove()
        snapshotObservations.removeValue(forKey: sessionID)
    }

    func startObservingRequests() {
        guard FirebaseApp.app() != nil, requestObservation == nil else { return }
        guard let uid = Auth.auth().currentUser?.uid else { return }
        let ref = firestoreProvider()
            .collection("users").document(uid)
            .collection("rollback_requests")
            // `inFlight` is the legacy pre-snake_case wire value; include it so a
            // doc still carrying it isn't excluded by Firestore's exact-match `in`
            // before the tolerant `Status(wireValue:)` decoder can run. Mirrors the
            // tolerance in RollbackContracts.Status and Android RollbackService.kt
            // (`whereIn("status", listOf("pending", "in_flight", "inFlight"))`).
            .whereField("status", in: ["pending", "in_flight", "inFlight"])
        requestObservation = ref.addSnapshotListener { [weak self] snapshot, _ in
            Task { @MainActor in
                guard let self else { return }
                let documents = snapshot?.documents ?? []
                let resolvedKey = try? await MobileCloudVaultKeyAccess.keyForReading(
                    uid: uid,
                    firestore: self.firestoreProvider()
                )
                let trustedKeys: [String: Data] = if let identity = resolvedKey?.signalIdentity {
                    await MobileCloudVaultSignalPayloads.trustedSenderPublicKeys(
                        uid: uid,
                        firestore: self.firestoreProvider(),
                        localIdentity: identity
                    )
                } else { [:] }
                let parsed: [RollbackRequest] = documents.compactMap { doc in
                    let signalData = try? resolvedKey.flatMap { key in
                        try MobileCloudVaultSignalPayloads.openSignalPayloadIfPresent(
                            doc.data(),
                            uid: uid,
                            collection: "rollback_requests",
                            docId: doc.documentID,
                            signalIdentity: key.signalIdentity,
                            trustedSenderPublicKeys: trustedKeys
                        )
                    }
                    return Self.decodeRequest(
                        data: doc.data() as NSDictionary,
                        documentID: doc.documentID,
                        vaultKey: resolvedKey?.keyData,
                        signalPlaintext: signalData ?? nil
                    )
                }
                self.pendingRequests = parsed
            }
        }
    }

    /// Submit a rollback request. Mac claims and applies; the phone watches
    /// `pendingRequests` for status transitions.
    @discardableResult
    func submit(sessionID: String, scope: RollbackScope, requestedBy: String) async throws -> RollbackRequest {
        guard FirebaseApp.app() != nil else { throw RollbackError.firebaseUnavailable }
        guard let uid = Auth.auth().currentUser?.uid else { throw RollbackError.notSignedIn }
        let request = RollbackRequest(sessionID: sessionID, scope: scope, requestedBy: requestedBy)
        let resolvedKey = try await MobileCloudVaultKeyAccess
            .keyForWriting(uid: uid, firestore: firestoreProvider())
        let ref = firestoreProvider()
            .collection("users").document(uid)
            .collection("rollback_requests").document(request.id)
        var payload = try Self.encodeRequest(request, vaultKey: resolvedKey.keyData)
        let signalState = MobileCloudVaultSignalPayloads.signalActivationState(domainID: "conversations_chat")
        if signalState != .off {
            let scopeData = try JSONEncoder().encode(request.scope)
            let privatePayload: NSDictionary = [
                "id": request.id,
                "sessionID": request.sessionID,
                "scopeJSON": String(data: scopeData, encoding: .utf8) ?? "{}",
                "requestedAt": ISO8601DateFormatter().string(from: request.requestedAt),
                "requestedBy": request.requestedBy,
                "status": request.status.rawValue,
                "schemaVersion": 1
            ]
            let plaintext = try JSONSerialization.data(withJSONObject: privatePayload, options: [.sortedKeys])
            do {
                if let envelope = try await MobileCloudVaultSignalPayloads.signalEnvelopeIfEnabled(
                    domainID: "conversations_chat",
                    uid: uid,
                    firestore: firestoreProvider(),
                    collection: "rollback_requests",
                    docId: request.id,
                    plaintext: plaintext,
                    resolvedKey: resolvedKey
                ) {
                    payload["signalEnvelope"] = envelope
                    if signalState == .required {
                        payload.removeValue(forKey: "sealedScope")
                        payload.removeValue(forKey: "sealedErrorMessage")
                        payload.removeValue(forKey: "sealedPayload")
                    }
                }
            } catch {
                if signalState == .required { throw error }
            }
        }
        try MobileCloudVaultSignalPayloads.requireEnvelopeIfRequired(
            payload: payload,
            state: signalState,
            domainID: "conversations_chat"
        )
        try await ref.setData(payload)
        return request
    }

    // MARK: - Decoding

    static func decodeSnapshot(
        data: NSDictionary,
        documentID: String,
        sessionID: String,
        vaultKey: Data?
    ) -> RollbackSnapshot? {
        guard
            let sequence = data["sequence"] as? Int,
            let takenAtString = data["takenAt"] as? String,
            let takenAt = ISO8601DateFormatter().date(from: takenAtString)
        else { return nil }
        // Open the sealed action label, falling back to legacy plaintext for
        // pre-migration documents. The label is required for the row to render.
        guard let actionLabel = openSealedString(
            data: data,
            sealedField: "sealedActionLabel",
            legacyField: "actionLabel",
            vaultKey: vaultKey
        ) else { return nil }
        // `touchedFiles` is sealed as one JSON array; legacy docs carried the
        // raw `[String]`.
        let touched = openSealedStringArray(
            data: data,
            sealedField: "sealedTouchedFiles",
            legacyField: "touchedFiles",
            vaultKey: vaultKey
        )
        let macPath = openSealedString(
            data: data,
            sealedField: "sealedMacSnapshotPath",
            legacyField: "macSnapshotPath",
            vaultKey: vaultKey
        )
        let restoredAt = (data["restoredAt"] as? String).flatMap { ISO8601DateFormatter().date(from: $0) }
        return RollbackSnapshot(
            id: (data["id"] as? String) ?? documentID,
            sessionID: sessionID,
            sequence: sequence,
            takenAt: takenAt,
            actionLabel: actionLabel,
            touchedFiles: touched,
            macSnapshotPath: macPath,
            restoredAt: restoredAt
        )
    }

    static func decodeRequest(
        data: NSDictionary,
        documentID: String,
        vaultKey: Data?,
        signalPlaintext: Data? = nil
    ) -> RollbackRequest? {
        var signal: NSDictionary = [:]
        if let signalPlaintext,
           let object = try? JSONSerialization.jsonObject(with: signalPlaintext),
           let dictionary = object as? NSDictionary {
            signal = dictionary
        } else if MobileCloudVaultSignalPayloads.signalSealingIsRequired(domainID: "conversations_chat") {
            return nil
        }
        let source = signal.count == 0 ? data : signal
        guard
            let sessionID = source["sessionID"] as? String,
            let scopeRaw = openSealedString(
                data: signal.count == 0 ? data : NSDictionary(dictionary: ["scopeJSON": source["scopeJSON"] as Any]),
                sealedField: "sealedScope",
                legacyField: "scopeJSON",
                vaultKey: vaultKey
            ),
            let scope = try? JSONDecoder().decode(RollbackScope.self, from: Data(scopeRaw.utf8)),
            // MUTABLE rollback state lives on the LIVE document: the Mac claim
            // path updates top-level `status`, `resolvedAt`, and
            // `sealedErrorMessage` without resealing the Signal envelope (the
            // envelope is created once at request time with `status: pending`).
            // Prefer the live fields so an `in_flight`/resolved request never
            // decodes as pending off the stale envelope; the Signal payload is
            // authoritative only for the private immutable fields and is the
            // fallback for docs the Mac has not touched yet.
            let statusRaw = (data["status"] as? String) ?? (source["status"] as? String),
            // Route through the tolerant resolver so the legacy camelCase
            // `"inFlight"` wire value (and `cancelled`) decode; a genuinely
            // unknown status still drops the row via the guard.
            let status = RollbackRequest.Status(wireValue: statusRaw)
        else { return nil }
        let requestedAt = (source["requestedAt"] as? String).flatMap { ISO8601DateFormatter().date(from: $0) } ?? Date()
        let resolvedAt = ((data["resolvedAt"] as? String) ?? (source["resolvedAt"] as? String))
            .flatMap { ISO8601DateFormatter().date(from: $0) }
        let requestedBy = (source["requestedBy"] as? String) ?? "unknown"
        let errorMessage = openSealedString(
            data: data,
            sealedField: "sealedErrorMessage",
            legacyField: "errorMessage",
            vaultKey: vaultKey
        ) ?? (signal["errorMessage"] as? String)
        return RollbackRequest(
            id: documentID,
            sessionID: sessionID,
            scope: scope,
            requestedAt: requestedAt,
            requestedBy: requestedBy,
            status: status,
            resolvedAt: resolvedAt,
            errorMessage: errorMessage
        )
    }

    static func encodeRequest(_ request: RollbackRequest, vaultKey: Data) throws -> [String: Any] {
        let scopeData = try JSONEncoder().encode(request.scope)
        let scopeJSON = String(data: scopeData, encoding: .utf8) ?? "{}"
        // Seal the scope JSON (a `singleFile` scope embeds an absolute file
        // path) instead of writing it in clear. The phone never writes
        // `errorMessage` — that is the Mac claim path's job and it seals
        // `sealedErrorMessage` from day one.
        return [
            "id": request.id,
            "sessionID": request.sessionID,
            "sealedScope": try dictionary(CloudVaultCrypto.sealText(scopeJSON, keyData: vaultKey)),
            "requestedAt": ISO8601DateFormatter().string(from: request.requestedAt),
            "requestedBy": request.requestedBy,
            "status": request.status.rawValue,
            "schemaVersion": 1,
            "source": "ios-hermes-square"
        ]
    }

    // MARK: - Seal helpers

    /// Opens a sealed-text field, falling back to a legacy plaintext field only
    /// when the sealed field is absent. A present sealed field is authoritative:
    /// decrypt it or fail closed instead of leaking a stale plaintext sibling.
    private static func openSealedString(
        data: NSDictionary,
        sealedField: String,
        legacyField: String,
        vaultKey: Data?
    ) -> String? {
        if let envelope = sealedText(from: data[sealedField]) {
            guard let vaultKey else { return nil }
            return try? CloudVaultCrypto.openText(envelope, keyData: vaultKey)
        }
        return data[legacyField] as? String
    }

    /// Opens a sealed `[String]` (sealed as one JSON array), falling back to a
    /// legacy plaintext `[String]` field only when the sealed field is absent.
    private static func openSealedStringArray(
        data: NSDictionary,
        sealedField: String,
        legacyField: String,
        vaultKey: Data?
    ) -> [String] {
        if sealedText(from: data[sealedField]) != nil {
            guard let json = openSealedString(
                data: data,
                sealedField: sealedField,
                legacyField: "",
                vaultKey: vaultKey
            ), let decoded = try? JSONDecoder().decode([String].self, from: Data(json.utf8)) else {
                return []
            }
            return decoded
        }
        return (data[legacyField] as? [String]) ?? []
    }

    /// Encodes a `CloudVaultSealedText` envelope into a Firestore-compatible
    /// dictionary. Mirrors the adjacent sealed-text sync services.
    private static func dictionary<T: Encodable>(_ value: T) throws -> [String: Any] {
        let data = try JSONEncoder().encode(value)
        let object = try JSONSerialization.jsonObject(with: data)
        guard let dictionary = object as? [String: Any] else { return [:] }
        return dictionary
    }

    /// Decodes a stored sealed-text dictionary back into a `CloudVaultSealedText`.
    private static func sealedText(from value: Any?) -> CloudVaultSealedText? {
        guard let dictionary = value as? [String: Any],
              let data = try? JSONSerialization.data(withJSONObject: dictionary) else {
            return nil
        }
        return try? JSONDecoder().decode(CloudVaultSealedText.self, from: data)
    }

    enum RollbackError: LocalizedError {
        case firebaseUnavailable
        case notSignedIn

        var errorDescription: String? {
            switch self {
            case .firebaseUnavailable: return "Firebase is not configured."
            case .notSignedIn:         return "Sign in to submit rollback requests."
            }
        }
    }
}

File: android/app/src/main/java/com/openburnbar/data/missions/RollbackService.kt
package com.openburnbar.data.missions

import com.google.firebase.FirebaseException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.openburnbar.data.cloud.AndroidCloudVaultKeyAccess
import com.openburnbar.data.cloud.CloudVaultCrypto
import com.openburnbar.data.firebase.CloudVaultSealedTextCodec
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

// MARK: - Rollback Service (Android parity, Hermes Square §6.10)
//
// Mirrors the iOS `RollbackService`: listens to
// `users/{uid}/cli_sessions/{sessionID}/snapshots`, exposes a
// `StateFlow<Map<String, List<RollbackSnapshot>>>`, and offers
// `submit(...)` which writes a rollback request to
// `users/{uid}/rollback_requests/{id}` for the Mac to claim.

data class RollbackSnapshot(
    val id: String,
    val sessionID: String,
    val sequence: Int,
    val takenAtEpoch: Long,
    val actionLabel: String,
    val touchedFiles: List<String>,
    val macSnapshotPath: String?,
    val restoredAtEpoch: Long?,
)

sealed class RollbackScope {
    object FullSession : RollbackScope()

    data class LastN(val count: Int) : RollbackScope()

    data class SingleFile(val path: String) : RollbackScope()

    val token: String
        get() =
            when (this) {
                FullSession -> "full_session"
                is LastN -> "last_$count"
                is SingleFile -> "file"
            }

    val asJson: String
        get() =
            when (this) {
                FullSession -> "{\"kind\":\"fullSession\"}"
                is LastN -> "{\"kind\":\"lastN\",\"count\":$count}"
                is SingleFile -> "{\"kind\":\"singleFile\",\"path\":${com.openburnbar.data.missions.jsonString(path)}}"
            }
}

data class RollbackRequest(
    val id: String,
    val sessionID: String,
    val scope: RollbackScope,
    val requestedAtEpoch: Long,
    val requestedBy: String,
    val status: Status,
    val resolvedAtEpoch: Long?,
    val errorMessage: String?,
) {
    enum class Status(val token: String) {
        PENDING("pending"),
        IN_FLIGHT("in_flight"),
        COMPLETED("completed"),
        FAILED("failed"),
        CANCELLED("cancelled"),
        ;

        companion object {
            fun fromToken(token: String?): Status? = when (token) {
                "inFlight" -> IN_FLIGHT
                null -> null
                else -> values().firstOrNull { it.token == token }
            }
        }
    }
}

class RollbackService private constructor(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(),
) {
    private val _snapshotsBySession = MutableStateFlow<Map<String, List<RollbackSnapshot>>>(emptyMap())
    val snapshotsBySession: StateFlow<Map<String, List<RollbackSnapshot>>> = _snapshotsBySession.asStateFlow()

    private val _pendingRequests = MutableStateFlow<List<RollbackRequest>>(emptyList())
    val pendingRequests: StateFlow<List<RollbackRequest>> = _pendingRequests.asStateFlow()

    private val _inlineError = MutableStateFlow<String?>(null)
    val inlineError: StateFlow<String?> = _inlineError.asStateFlow()

    private val snapshotRegistrations = ConcurrentHashMap<String, ListenerRegistration>()
    private var requestsRegistration: ListenerRegistration? = null
    private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun startObservingSession(sessionID: String) {
        if (snapshotRegistrations.containsKey(sessionID)) return
        val uid = auth.currentUser?.uid ?: return
        // Reserve the slot synchronously so concurrent calls don't double-register
        // while the suspend read-key resolves.
        snapshotRegistrations[sessionID] = NoopListenerRegistration
        ioScope.launch {
            // Snapshots seal `sealedActionLabel`/`sealedTouchedFiles`/
            // `sealedMacSnapshotPath` (Mac claimed-device writer). Resolve the read
            // key once so the decoder can open them; legacy plaintext snapshots fall
            // back inside `toRollbackSnapshotOrNull`.
            val vaultKey =
                runCatching {
                    AndroidCloudVaultKeyAccess.keyForReading(uid = uid, firestore = firestore)?.keyData
                }.getOrNull()
            val ref =
                firestore.collection("users").document(uid)
                    .collection("cli_sessions").document(sessionID)
                    .collection("snapshots")
                    .orderBy("sequence")
            val reg =
                ref.addSnapshotListener { snap, error ->
                    if (error != null) {
                        _inlineError.value = error.localizedMessage
                        return@addSnapshotListener
                    }
                    val parsed =
                        snap?.documents.orEmpty().mapNotNull { doc ->
                            doc.data?.toRollbackSnapshotOrNull(documentID = doc.id, sessionID = sessionID, vaultKey = vaultKey)
                        }
                    _snapshotsBySession.value = _snapshotsBySession.value + (sessionID to parsed)
                }
            // Stop-during-resolve removes the placeholder; honor that by dropping
            // the freshly attached listener instead of leaking it.
            if (!snapshotRegistrations.replace(sessionID, NoopListenerRegistration, reg)) {
                reg.remove()
            }
        }
    }

    fun stopObservingSession(sessionID: String) {
        snapshotRegistrations.remove(sessionID)?.remove()
        _snapshotsBySession.value = _snapshotsBySession.value - sessionID
    }

    fun startObservingRequests() {
        if (requestsRegistration != null) return
        val uid = auth.currentUser?.uid ?: return
        // Reserve synchronously to keep the idempotent guard above honest while the
        // suspend read-key resolves.
        requestsRegistration = NoopListenerRegistration
        ioScope.launch {
            // Requests seal `sealedScope` (and, when the Mac claim path lands,
            // `sealedErrorMessage`). Resolve the read key so the decoder can open
            // them; legacy plaintext requests fall back inside
            // `toRollbackRequestOrNull`.
            val vaultKey =
                runCatching {
                    AndroidCloudVaultKeyAccess.keyForReading(uid = uid, firestore = firestore)?.keyData
                }.getOrNull()
            val ref =
                firestore.collection("users").document(uid)
                    .collection("rollback_requests")
                    .whereIn("status", listOf("pending", "in_flight", "inFlight"))
            val reg =
                ref.addSnapshotListener { snap, error ->
                    if (error != null) {
                        _inlineError.value = error.localizedMessage
                        return@addSnapshotListener
                    }
                    val parsed =
                        snap?.documents.orEmpty().mapNotNull { doc ->
                            doc.data?.toRollbackRequestOrNull(documentID = doc.id, vaultKey = vaultKey)
                        }
                    _pendingRequests.value = parsed
                }
            if (requestsRegistration === NoopListenerRegistration) {
                requestsRegistration = reg
            } else {
                // stopAll() ran during resolution — don't leak the listener.
                reg.remove()
            }
        }
    }

    fun stopAll() {
        snapshotRegistrations.values.forEach { it.remove() }
        snapshotRegistrations.clear()
        requestsRegistration?.remove()
        requestsRegistration = null
        _snapshotsBySession.value = emptyMap()
        _pendingRequests.value = emptyList()
    }

    suspend fun submit(sessionID: String, scope: RollbackScope, requestedBy: String): RollbackRequest? {
        val uid =
            auth.currentUser?.uid ?: run {
                _inlineError.value = "Sign in to submit rollback requests."
                return null
            }
        val id = UUID.randomUUID().toString()
        val now = Instant.now()
        val request =
            RollbackRequest(
                id = id,
                sessionID = sessionID,
                scope = scope,
                requestedAtEpoch = now.toEpochMilli(),
                requestedBy = requestedBy,
                status = RollbackRequest.Status.PENDING,
                resolvedAtEpoch = null,
                errorMessage = null,
            )
        // Seal the scope JSON (which embeds an absolute file path for `singleFile`
        // scope) with the Cloud Vault key, mirroring the iOS writer. Do not fall
        // back to plaintext when this device is not approved; keep the request
        // local and make the user approve the device before cloud sync.
        val vaultKey =
            runCatching {
                AndroidCloudVaultKeyAccess.keyForWriting(uid = uid, firestore = firestore).keyData
            }.getOrElse { error ->
                _inlineError.value = error.localizedMessage ?: "Approve this Android device before syncing rollback requests."
                return null
            }
        val payload = sealedRollbackRequestPayload(request, source = "android-hermes-square", vaultKey = vaultKey)
        return try {
            firestore.collection("users").document(uid)
                .collection("rollback_requests").document(id)
                .set(payload)
                .await()
            request
        } catch (e: FirebaseException) {
            _inlineError.value = e.localizedMessage ?: "Rollback request failed."
            null
        }
    }

    companion object {
        @Volatile private var instance: RollbackService? = null

        fun shared(): RollbackService = instance ?: synchronized(this) {
            instance ?: RollbackService().also { instance = it }
        }
    }
}

internal fun sealedRollbackRequestPayload(request: RollbackRequest, source: String, vaultKey: ByteArray): MutableMap<String, Any> = mutableMapOf(
    "id" to request.id,
    "sessionID" to request.sessionID,
    "requestedAt" to Instant.ofEpochMilli(request.requestedAtEpoch).toString(),
    "requestedBy" to request.requestedBy,
    "status" to request.status.token,
    "schemaVersion" to 1,
    "source" to source,
    "sealedScope" to CloudVaultSealedTextCodec.toMap(CloudVaultCrypto.sealText(request.scope.asJson, vaultKey)),
)

internal fun Map<String, Any?>.toRollbackSnapshotOrNull(documentID: String, sessionID: String, vaultKey: ByteArray? = null): RollbackSnapshot? {
    val sequence = (this["sequence"] as? Number)?.toInt()
    val takenAtIso = this["takenAt"] as? String
    val takenAtEpoch = takenAtIso?.let { runCatching { Instant.parse(it).toEpochMilli() }.getOrNull() }
    val actionLabel =
        CloudVaultSealedTextCodec.openOrLegacy(this["sealedActionLabel"], vaultKey, this["actionLabel"] as? String)
    if (sequence == null || takenAtEpoch == null || actionLabel == null) return null
    // `sealedTouchedFiles` seals the whole `[String]` array as one JSON string.
    // If the sealed field exists, fail closed instead of falling back to a stale
    // plaintext sibling.
    val touched =
        openSealedStringListOrLegacy(this["sealedTouchedFiles"], vaultKey, this["touchedFiles"])
    val macSnapshotPath =
        CloudVaultSealedTextCodec.openOrLegacy(
            this["sealedMacSnapshotPath"],
            vaultKey,
            this["macSnapshotPath"] as? String,
        )
    val restoredAtIso = this["restoredAt"] as? String
    val restoredAtEpoch = restoredAtIso?.let { runCatching { Instant.parse(it).toEpochMilli() }.getOrNull() }
    return RollbackSnapshot(
        id = this["id"] as? String ?: documentID,
        sessionID = sessionID,
        sequence = sequence,
        takenAtEpoch = takenAtEpoch,
        actionLabel = actionLabel,
        touchedFiles = touched,
        macSnapshotPath = macSnapshotPath,
        restoredAtEpoch = restoredAtEpoch,
    )
}

internal fun Map<String, Any?>.toRollbackRequestOrNull(documentID: String, vaultKey: ByteArray? = null): RollbackRequest? {
    val sessionID = this["sessionID"] as? String ?: return null
    val scopeJSON =
        CloudVaultSealedTextCodec.openOrLegacy(this["sealedScope"], vaultKey, this["scopeJSON"] as? String)
            ?: return null
    val scope = parseScope(scopeJSON)
    val statusRaw = this["status"] as? String
    val status = RollbackRequest.Status.fromToken(statusRaw) ?: return null
    val requestedAt =
        (this["requestedAt"] as? String)?.let {
            runCatching { Instant.parse(it).toEpochMilli() }.getOrNull()
        } ?: System.currentTimeMillis()
    val resolvedAt =
        (this["resolvedAt"] as? String)?.let {
            runCatching { Instant.parse(it).toEpochMilli() }.getOrNull()
        }
    val requestedBy = this["requestedBy"] as? String ?: "unknown"
    val errorMessage =
        CloudVaultSealedTextCodec.openOrLegacy(this["sealedErrorMessage"], vaultKey, this["errorMessage"] as? String)
    return RollbackRequest(
        id = documentID,
        sessionID = sessionID,
        scope = scope,
        requestedAtEpoch = requestedAt,
        requestedBy = requestedBy,
        status = status,
        resolvedAtEpoch = resolvedAt,
        errorMessage = errorMessage,
    )
}

private fun openSealedStringListOrLegacy(rawSealed: Any?, vaultKey: ByteArray?, rawLegacy: Any?): List<String> {
    if (CloudVaultSealedTextCodec.fromMap(rawSealed as? Map<*, *>) != null) {
        val json = CloudVaultSealedTextCodec.open(rawSealed, vaultKey) ?: return emptyList()
        return runCatching {
            val arr = org.json.JSONArray(json)
            (0 until arr.length()).map { arr.getString(it) }
        }.getOrNull() ?: emptyList()
    }
    return (rawLegacy as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
}

private fun parseScope(json: String): RollbackScope {
    return runCatching {
        val obj = org.json.JSONObject(json)
        when (obj.optString("kind")) {
            "fullSession" -> RollbackScope.FullSession
            "lastN" -> RollbackScope.LastN(obj.optInt("count", 1))
            "singleFile" -> RollbackScope.SingleFile(obj.optString("path"))
            else -> RollbackScope.FullSession
        }
    }.getOrDefault(RollbackScope.FullSession)
}

// Placeholder reservation so the synchronous `startObserving*` idempotency guard
// stays honest while the suspend Cloud Vault read key resolves on `ioScope`.
private object NoopListenerRegistration : ListenerRegistration {
    override fun remove() = Unit
}

internal fun jsonString(raw: String): String {
    val out = StringBuilder("\"")
    for (ch in raw) {
        when (ch) {
            '"' -> out.append("\\\"")
            '\\' -> out.append("\\\\")
            '\b' -> out.append("\\b")
            '\n' -> out.append("\\n")
            '\r' -> out.append("\\r")
            '\t' -> out.append("\\t")
            else -> if (ch < ' ') out.append("\\u%04x".format(ch.code)) else out.append(ch)
        }
    }
    out.append('"')
    return out.toString()
}
